### PR TITLE
AGDIGGER-56 - change logging endpoint to use build number

### PR DESCRIPTION
**Pull request by wtrocki**
## Motivation

Align node.js implementation to use build number instead of queue number as in java client.

### CODE REVIEW

**Comment by wtrocki** [comment]
ping @aliok @luigizuccarelli @odra

**Comment by wtrocki** [comment]
Review @josemigallas

File: lib/api/triggerBuild.js [hunk 1]
```diff
@@ -8,9 +9,22 @@ module.exports = (auth, jobName, callback) => {
...
+    if (err) {
+      callback(err);
+      return;
+    }
+    logger.info("Build triggered. Waiting for build number.");
+    const fetchInterval = setInterval(() => {
+      jenkins.queue.item(queueNumber, function(err, data) {
+        if (data && data.executable && data.executable.number) {
+          clearInterval(fetchInterval);
+          callback(null,data.executable.number);
```
Comment (josemigallas) [inline lib/api/triggerBuild.js]: Here is missing one space after `null, `

File: test/api/createJobTests.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+const assert = chai.assert;
+
+const auth = {
+    url: "test",
+    user: "user",
+    password: "password"
+};
+
+const jobArgs = {
+        repository: "test",
```
Comment (josemigallas) [inline test/api/createJobTests.js]: Weird indentation here

File: test/api/createJobTests.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+const dataMock = {name:"Test"};
+
+const JenkinsWrapper = {
+    job: {
+        create: sinon.stub().yields(null, dataMock)
+    }
+};
+
+const jenkinsStub = () =>
+    JenkinsWrapper;
```
Comment (josemigallas) [inline test/api/createJobTests.js]: It would look nicer to have all in the same line `const jenkinsStub = () => JenkinsWrapper;`

File: test/api/createJobTests.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    password: "password"
+};
+
+const jobArgs = {
+        repository: "test",
+        branch: "master",
+        name: "test"
+    }
+
+const dataMock = {name:"Test"};
```
Comment (josemigallas) [inline test/api/createJobTests.js]: I would put this like the other constants for consistency, in JSONish form.
```
const dataMock = {
    name:"Test"
};
```

File: test/api/createJobTests.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    JenkinsWrapper;
+
+const createJob = proxyquire("../../lib/api/createJob", { 'jenkins': jenkinsStub });
+
+describe('createJob', function () {
+    it('it should return job', function () {
+        createJob(auth, jobArgs, function (err, data) {
+            assert.isNull(err, "Should not return error")
+            assert.isNotNull(data, "Should return data");
+             assert.equal(data.name,dataMock.name );
```
Comment (josemigallas) [inline test/api/createJobTests.js]: This line is one space misplaced and misses one space after `data.name,  `.

File: test/api/createJobTests.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+const jenkinsStub = () =>
+    JenkinsWrapper;
+
+const createJob = proxyquire("../../lib/api/createJob", { 'jenkins': jenkinsStub });
+
+describe('createJob', function () {
+    it('it should return job', function () {
+        createJob(auth, jobArgs, function (err, data) {
+            assert.isNull(err, "Should not return error")
```
Comment (josemigallas) [inline test/api/createJobTests.js]: Could it happen that the error is not null but `undefined`? Should we maybe use `isNotOk()` instead?

File: test/api/fetchArtifactTests.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+const chai = require("chai");
+const assert = chai.assert;
+
+const auth = {
+  url: "test",
+  user: "user",
+  password: "password"
+};
+
+const dataMock = {artifacts:[{relativePath:"test"}]};
```
Comment (josemigallas) [inline test/api/fetchArtifactTests.js]: I would use the extended, JSON-like form.

File: test/api/fetchArtifactTests.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+const JenkinsWrapper = {
+    build: {
+        get: sinon.stub().yields(null, dataMock)
+    }
+};
+
+const jenkinsStub = () =>
+     JenkinsWrapper;
```
Comment (josemigallas) [inline test/api/fetchArtifactTests.js]: It would look nicer in one line

File: test/api/fetchArtifactTests.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+var fetchArtifact = proxyquire("../../lib/api/fetchArtifact", { 'jenkins': jenkinsStub });
+
+describe('fetchArtifact', function() {
+  it('it should return artifacts', function() {
+    fetchArtifact(auth,"job", 1 , function(err, data) {
```
Comment (josemigallas) [inline test/api/fetchArtifactTests.js]: A espace is missing after `auth, `.

File: test/api/fetchArtifactTests.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+      assert.isNull(err);
```
Comment (josemigallas) [inline test/api/fetchArtifactTests.js]: Again, what if err is not null but undefined?

File: test/api/fetchArtifactTests.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+      assert.isNotNull(data);
+      assert.equal(data.length,1);
+      assert.equal(data[0],"test/job/job/1/artifact/test");
```
Comment (josemigallas) [inline test/api/fetchArtifactTests.js]: Missing a espace after `data[0], `. Moreover, should this string not be hardcoded but in a const string?

File: test/api/jenkinsInfoTests.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+const jenkinsStub = () =>
+     JenkinsWrapper;
+
+var jenkinsInfo = proxyquire("../../lib/api/jenkinsInfo", { 'jenkins': jenkinsStub });
+
+describe('jenkinsInfo', function() {
+  it('it should return jenkinsInfo', function () {
+    jenkinsInfo(auth, function (err, data) {
+      assert.isNull(err, "Should not return error")
```
Comment (josemigallas) [inline test/api/jenkinsInfoTests.js]: Again as above, what if err is not null but undefined?

File: test/api/streamLogsTests.js [hunk 13]
```diff
@@ -1,39 +1,47 @@
...
+const JenkinsWrapper = {
+  queue: {
+    item: sinon.stub().yields(null, data)
+  },
+  build: {
+    logStream: () => ({ on: function() { } })
   }
 };
 
+const jenkinsStub = () => JenkinsWrapper;
```
Comment (josemigallas) [inline test/api/streamLogsTests.js:30]: Finally! It is in one line here 😄

File: test/api/streamLogsTests.js [hunk 14]
```diff
@@ -1,39 +1,47 @@
...
-  it('it should stream logs', function () {
-    var auth = {}
-    streamLogs(auth, jobName, 1, console, function (err) {
-      assert(!err);
-    })
+describe('Streaming logs', function() {
+  it('it should return error', function() {
+    JenkinsWrapper.queue.item = sinon.stub().yields("error");
+    streamLogs(auth, jobName, 1, console, function(err) {
+      assert.isNull(err);
```
Comment (josemigallas) [inline test/api/streamLogsTests.js]: Again as above and also the test below, what if not null but undefined?

File: test/api/triggerBuildTests.js [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+  password: "password"
+};
+
+
+const JenkinsWrapper = {
+    job: { build: sinon.stub().yields(null, {}) }
+};
+
+const jenkinsStub = () =>
+     JenkinsWrapper;
```
Comment (josemigallas) [inline test/api/triggerBuildTests.js]: One line looks nicer?

File: test/api/triggerBuildTests.js [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+var triggerBuild = proxyquire("../../lib/api/triggerBuild", { 'jenkins': jenkinsStub });
+
+describe('triggerBuild', function() {
+  it('it should trigger build', function () {
+    triggerBuild(auth, "job", function (err, data) {
+      assert.isNull(err, "Should not return error")
```
Comment (josemigallas) [inline test/api/triggerBuildTests.js]: Again as above, what if not null but indefined?

File: test/api/triggerBuildTests.js [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+      assert.isNotNull(data, "Should return data");
```
Comment (josemigallas) [inline test/api/triggerBuildTests.js]: Here happens the opposite. If data happens to be `undefined` this test will pass. It should be `.isOk()`.

**Review by josemigallas — COMMENTED** [review]
I've seen some format issues but the important are the issues about isNull and isNotNull

File: test/api/createJobTests.js (hunk 6 above)
Comment (wtrocki) [reply test/api/createJobTests.js]: Great comment!

**Comment by matskiv** [comment]
@josemigallas has some good comments, otherwise LGTM and works well. :+1:

File: README.md [hunk 18]
```diff
@@ -51,7 +51,7 @@ digkins job build my-androip-build
   digkins login <url> [user] [password] 
 
   Stream jenkins logs for triggered build
-  digkins log <job> <queueid>       
+  digkins log <job> <buildNumber>
```
Comment (matzew) [inline README.md]: is  that the real build number from Jenkins? Or this other number we had before, which did not match the actual Jenkins Build NR

File: package.json [hunk 19]
```diff
@@ -40,6 +40,7 @@
     "node": ">= 4"
   },
   "devDependencies": {
+    "chai": "^3.5.0",
```
Comment (matzew) [inline package.json]: makes sense to use fixed versions ? Instead of ranges ?

File: test/api/fetchArtifactTests.js (hunk 7 above)
Comment (matzew) [reply test/api/fetchArtifactTests.js]: +1

File: test/api/triggerBuildTests.js (hunk 15 above)
Comment (matzew) [reply test/api/triggerBuildTests.js]: +1 I agree

File: lib/api/streamLogs.js [hunk 20]
```diff
@@ -4,38 +4,25 @@ const Jenkins = require('jenkins');
...
-};
-
 
+  logStream.on('error', function(err) {
+    if (err) {
+      writter.error(err);
+    }
+  });
 
+  logStream.on('end', function(err) {
```
Comment (odra) [inline lib/api/streamLogs.js]: I usually prefer to use the new callback syntax for callbacks `() => {code}`

File: test/api/fetchArtifactTests.js (hunk 8 above)
Comment (odra) [reply test/api/fetchArtifactTests.js]: I think it makes sense to use `= function()...` instead of the new syntax (since it is not a callback function)

File: lib/api/streamLogs.js (hunk 20 above)
Comment (josemigallas) [reply lib/api/streamLogs.js]: +100

File: package.json (hunk 19 above)
Comment (josemigallas) [reply package.json]: I think the recommended way is to use fixed versions. That way if anything that would break your code changes in a new version, you are safe.

Comment (odra) [reply package.json]: +1

**Comment by odra** [comment]
Seems good, added some comments.

Also, I think it would be a good idea to add eslint verification in travis.yaml file (in another PR).

File: test/api/fetchArtifactTests.js (hunk 8 above)
Comment (josemigallas) [reply test/api/fetchArtifactTests.js]: Here is a matter of cleanness (and personal preference 😂 )
```
const jenkinsStub = () => JenkinsWrapper;
```
Reads better than
```
const jenkinsStub = function() {
    return JenkinsWrapper;
}
```

Comment (odra) [reply test/api/fetchArtifactTests.js]: why not just `const jenkinsStub = JenkinsWrapper`? (or use let/var if const causes some issue here)

File: lib/api/streamLogs.js (hunk 20 above)
Comment (wtrocki) [reply lib/api/streamLogs.js]: I use them as well, but sometimes just blindly copying and pasting snippets from internet and forgetting to change - you know this ugly copy and paste coding style ;)

File: package.json (hunk 19 above)
Comment (wtrocki) [reply package.json]: Not this discussion again ;) There is only one right answer here: 
**"We should use shrinkwrap to lock versions"**
....



**More seriously now:**
Fully agree going to lock down all dependencies, but I needed to mention shrinkwrap here just for fun!

File: test/api/fetchArtifactTests.js (hunk 8 above)
Comment (josemigallas) [reply test/api/fetchArtifactTests.js]: Yeah that's true! @wtrocki Why is this even a function?

**Comment by wtrocki** [comment]
@odra eslint is already integrated into the travis.

**Comment by wtrocki** [comment]
## Verification steps
1. Create job
2. Trigger build
3. Fetch logs

```
digkins login jenkins.localhost.io:8080 admin Password1                                                           
digkins job create demo https://github.com/wtrocki/helloworld-android-gradle master                               
digkins job build demo                                                                                           
digkins log demo 1                                                                                               
digkins artifact demo 1     
```
